sw/device/tests/edn_kat.c

```diff
@@ -0,0 +1,213 @@
+                  kEdnKatSeedMaterialGenerate.len,
+                  /*generate_len=*/
+                  alert_test ? 1 : kEdnKatOutputLen / kEdnKatWordsPerBlock),
+              .seed_material = kEdnKatSeedMaterialGenerate,
+          },
+      .reseed_interval = alert_test ? 1 : 32,
+  };
+}
+
+static uint32_t ibex_get_rnd_data(void) {
+  uint32_t ibex_rnd_data;
+  CHECK_STATUS_OK(rv_core_ibex_testutils_get_rnd_data(
+      &rv_core_ibex, kEdnKatTimeout, &ibex_rnd_data));
+  return ibex_rnd_data;
+}
+
+// Configure the entropy complex.
+static void entropy_config(bool alert_test) {
+  dif_edn_auto_params_t edn_params =
+      edn_testutils_auto_params_build(true, /*res_itval=*/0, /*glen_val=*/0);
+  // Disable the entropy complex.
+  CHECK_STATUS_OK(entropy_testutils_stop_all());
+  // Enable ENTROPY_SRC in FIPS mode.
+  CHECK_DIF_OK(dif_entropy_src_configure(
+      &entropy_src, entropy_testutils_config_default(), kDifToggleEnabled));
+  // Enable CSRNG.
+  CHECK_DIF_OK(dif_csrng_configure(&csrng));
+  // Enable EDN1 in auto request mode.
+  CHECK_DIF_OK(dif_edn_set_auto_mode(&edn1, edn_params));
+  // Enable EDN0 in auto request mode.
+  CHECK_DIF_OK(dif_edn_set_auto_mode(&edn0, kat_auto_params_build(alert_test)));
+}
+
+static bool execute_kat(void) {
+  uint32_t ibex_rnd_data_got;
+  unsigned int next_val_pos = 0;
+  unsigned int hits = 0;
+
+  // kExpectedOutput contains kEdnKatOutputLen precalculated words of the
+  // expected random output. The EDN might hold some buffered words from
+  // previous instantiations of the entropy complex. Since there are other
+  // blocks that consume entropy in between ibex rnd data requests, the first
+  // hit might not be the first word in the expected sequence. The following
+  // hits might also skip some words in the expected sequence if random words
+  // have been stolen by other blocks.
+  while (next_val_pos < kEdnKatOutputLen) {
+    // Fetch a new word of random bits.
+    ibex_rnd_data_got = ibex_get_rnd_data();
+    // Iterate over the remaining words in the expected sequence.
+    for (unsigned int i = next_val_pos; i < kEdnKatOutputLen; i++) {
+      // If there is a hit, the next potential hit is the current position +1.
+      if (ibex_rnd_data_got == kExpectedOutput[i]) {
+        next_val_pos = i + 1;
+        hits++;
+        break;
+        // If the first hit has happened and there is no hit in this position,
+        // this word has been stolen by another block. Thus, we need to
+        // increment the potential next_val_pos.
+      } else if (hits) {
+        next_val_pos = i + 1;
+      }
+    }
+  }
+  // If less than half of the words in the expected sequence were hit throw an
+  // exception.
+  LOG_INFO("%d/%d expected random words have been hit.", hits,
+           kEdnKatOutputLen);
+  CHECK(hits >= kEdnKatOutputLen / 2);
+  return true;
+}
+
+static void execute_alert_test(void) {
+  uint32_t alerts;
+  for (int i = 0; i <= kEdnKatWordsPerBlock; i++) {
```

Comment:
I am not sure how reliable this is. Because this will cause Ibex to read exactly 4 words. This means EDN needs to at least fetch one 128-bit block from CSRNG. The repetition checks run on these 128-bit blocks (not the 32-bit ones delivered to consumers). Depending on clock ratios and how much entropy other consumers on EDN0 really consume while Ibex is doing these reads, I believe there is a real possibility that Ibex does the only for 4 requests to EDN0 during that time, meaning a single 128-bit block is fetched from CSRNG and we don't get the alert. The test will fail.

To be on the safe side, I would change this to read `2*kEdnKatWordsPerBlock`. Then you're absolutely sure that just due to the reads of Ibex, two 128-bit blocks will be fetch from CSRNG.

Reply:
This makes a lot of sense. I'll change the code accordingly in my next push. Thanks!